jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/annotations/InitialAnnotationResults.kt

```diff
@@ -0,0 +1,52 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+package software.aws.toolkits.jetbrains.services.cloudformation.annotations
+
+import com.intellij.json.JsonFileType
+import com.intellij.json.psi.JsonFile
+import com.intellij.json.psi.JsonObject
+import com.intellij.openapi.application.runReadAction
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import org.jetbrains.yaml.YAMLFileType
+import org.jetbrains.yaml.psi.YAMLFile
+import software.aws.toolkits.jetbrains.services.cloudformation.yaml.YamlCloudFormationTemplate
+
+class InitialAnnotationResults(psiFile: PsiFile) {
+
+    private val TEMPLATE_VERSION_KEY = "AWSTemplateFormatVersion"
+    private val RESOURCES_KEY = "Resources"
+
+    var isCloudFormationTemplate = false
+
+    val pathToTemplate = psiFile.virtualFile.path
+
+    init {
+        if (psiFile.fileType is JsonFileType) {
```

Comment:
Can we rebased this off [feature/cloudformation](https://github.com/aws/aws-toolkit-jetbrains/tree/feature/cloudformation) and use the newly introduced CFN FileTypes (https://github.com/aws/aws-toolkit-jetbrains/commit/3a60a54539686559c9819eb74fa88b6f57c028d9)?

Reply:
While `AWS::Serverless-2016-10-31` and `AWSTemplateFormatVersion` definitely indicate CloudFormation templates, I suspect most CloudFormation templates don't contain either

It'd be nice if other CloudFormation templates could be recognized using something like [`Resources` and `Type`](https://github.com/aws-cloudformation/aws-cfn-lint-visual-studio-code/blob/edd51fb86441a1c48d294c4b664189c6f4d11399/server/src/server.ts#L132-L133)

Reply:
Agree with keeping CloudFormation template recognition in one place though

Reply:
> It'd be nice if other CloudFormation templates could be recognized using something like Resources and Type

I included this in the latest commit

Reply:
see my commend above - I'm not sure this resolves it.

Reply:
@kiiadi, not sure how to proceed. The code I wrote handles CFN templates but not SAM templates. If @abrooksv incorporates pat's comment into this branch, I can definitely rebase on top of that. 

So the logic should look something like

```java
if (contains "Transform: AWS::Serverless-2016-10-31")
   return true
if (contains "AWSTemplateFormatVersion: "2010-09-09"")
   return true
if (contains "Resources" block AND each Resource has a "Type" key) 
   return true

return false
```

Reply:
The last case is too unstable to use. As a user is trying to add a new resource the file will "blink" between being a CFN file and not being a CFN file. This has repercussions of things like auto-complete.

We will probably allow people to mark arbitrary files as CFN files and cache that as a VirtualFile attribute, but I have not made that change yet.

Reply:
> The code I wrote handles CFN templates but not SAM templates

wouldn't hurt to add `Transform: AWS::Serverless-2016-10-31` since that definitely indicates a template, but [SAM templates](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/transform-aws-serverless.html) should already be detected by the existing logic

> ```java
>    if (psiFile contains "Resources" block AND each Resource has a "Type" key) 
> ```
> The last case is too unstable to use. As a user is trying to add a new resource the file will "blink" between being a CFN file and not being a CFN file. This has repercussions of things like auto-complete



don't think this necessarily needs to be checked for every resource. [The Visual Studio Code extension only checks for one instance](https://github.com/aws-cloudformation/aws-cfn-lint-visual-studio-code/blob/edd51fb86441a1c48d294c4b664189c6f4d11399/server/src/server.ts#L133) and seems to be sufficient:

```java
   if (psiFile contains "Resources" block AND a Resource has a "Type" key) 
```

Reply:
also might want to filter out https://www.serverless.com/ files like the VSCode extension does if we include the `Resources` and `Type` logic:
https://github.com/aws-cloudformation/aws-cfn-lint-visual-studio-code/pull/25#issuecomment-642416464

Reply:
> We will probably allow people to mark arbitrary files as CFN files and cache that as a VirtualFile attribute, but I have not made that change yet.

This doesn't sound very user-friendly.

Would it make sense to only look out for `"Transform: AWS::Serverless-2016-10-31"` and `"AWSTemplateFormatVersion: \"2010-09-09\""`? Plus introducing a new warning rule in `cfn-lint` that using these are good practices? (I don't understand why these were not made required tags from the very beginning)

Reply:
> Plus introducing a new warning rule in `cfn-lint` that using these are good practices? (I don't understand why these were not made required tags from the very beginning)

The [Linter does recommend `Transform: AWS::Serverless-2016-10-31`](https://github.com/aws-cloudformation/cfn-python-lint/blob/master/src/cfnlint/rules/resources/ServerlessTransform.py) for the [Serverless transform](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/transform-aws-serverless.html) if the template contains [`AWS::Serverless` resource types](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-specification-resources-and-properties.html)

`Transform: AWS::Serverless-2016-10-31` should definitely not be recommended for every CloudFormation template. It requires an additional capability, `CAPABILITY_AUTO_EXPAND`, to be explicitly granted for stack operations to continue succeeding, [Stack sets](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/what-is-cfnstacksets.html) [does not support transforms](https://github.com/aws-cloudformation/aws-cloudformation-coverage-roadmap/issues/207), and many other features like [`aws cloudformation validate-template` and `aws cloudformation get-template-summary` do not understand templates with transforms](https://i.amazon.com/CFN-DE-1693?selectedConversation=f0a390bc-664c-445a-b72f-752023e3738e)

Reply:
> > Plus introducing a new warning rule in `cfn-lint` that using these are good practices? (I don't understand why these were not made required tags from the very beginning)
> 
> The [Linter does recommend `Transform: AWS::Serverless-2016-10-31`](https://github.com/aws-cloudformation/cfn-python-lint/blob/master/src/cfnlint/rules/resources/ServerlessTransform.py) for the [Serverless transform](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/transform-aws-serverless.html) if the template contains [`AWS::Serverless` resource types](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-specification-resources-and-properties.html)

Ok, what if it also recommended adding `AWSTemplateFormatVersion: "2010-09-09"` to the templates so that in this plugin we only look out for that?

Reply:
> in this plugin we only look out for that?

We should still look out for more signals because [`AWSTemplateFormatVersion`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/format-version-structure.html) can never be required by CloudFormation itself at this point. No reason to discard the other signals even if we were to start encouraging one of them

Reply:
The API for detecting file types by content is limited to the first 1024 bytes, and you don't have access to a pre-parsed file. This is what is needed if you want to support `foo.template` files which the existing plugin does.

For when you know the file extension, you do get a PSI file for it, but you still need this action to be as quick as possible due to it will be called for every single file of that type and multiple times. You need to balance speed for accuracy

Reply:
@kiiadi @abrooksv what's the status of https://github.com/aws/aws-toolkit-jetbrains/tree/feature/cloudformation ?